distributed/core.py

```diff
@@ -1599,12 +1594,13 @@ def callback(task: asyncio.Task[Comm]) -> None:
         except asyncio.CancelledError:
             # This is an outside cancel attempt
             connect_attempt.cancel()
-            try:
-                await connect_attempt
-            except CommClosedError:
-                pass
+            await connect_attempt
             raise
-        return await connect_attempt
+        try:
+            return connect_attempt.result()
+        except asyncio.CancelledError:
+            reason = self._reasons.pop(connect_attempt, "ConnectionPool closing.")
+            raise CommClosedError(reason)
```

Comment:
@graingert I'm all ears if you know how to better test this

Reply:
I got `test_remove_cancels_connect_before_task_running` which appears to produce this condition 🎉  (not a perfect test but it's something)